lib/Scheduling/Scheduler.cpp

```diff
@@ -0,0 +1,156 @@
+//===- Scheduler.cpp - Common interface for scheduling algorithms -===========//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements base classes and utilities for scheduling algorithms.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Scheduling/Scheduler.h"
+
+#include "mlir/IR/Operation.h"
+
+using namespace circt;
+using namespace circt::sched;
+
+Scheduler::Scheduler(Operation *containingOp) : containingOp(containingOp) {
+  uniquer.registerParametricStorageType<Dependence>();
+  uniquer.registerParametricStorageType<OperatorType>();
+}
+
+Dependence *Scheduler::getOrInsertDependence(Operation *src, unsigned srcIdx,
+                                             Operation *dst, unsigned dstIdx) {
+  auto *dep = uniquer.get<Dependence>({}, src, dst, 0, 0);
+  deps.insert(dep);
+  return dep;
+}
+
+OperatorType *Scheduler::getOrInsertOperatorType(StringRef name) {
+  auto *opr = uniquer.get<OperatorType>({}, name);
+  oprs.insert(opr);
+  return opr;
+}
+
+LogicalResult Scheduler::checkOp(Operation *op) {
+  if (!hasAssociatedOperatorType(op)) {
+    op->emitError("Operation is not associated with an operator type");
+    return failure();
+  }
+  return success();
+}
+
+LogicalResult Scheduler::checkDep(Dependence *dep) {
+  Operation *i = dep->getSource();
+  Operation *j = dep->getDestination();
+  if (!(hasOperation(i) && hasOperation(j))) {
+    containingOp->emitError()
+        << "Scheduling problem contains dependence with unregistered endpoints."
+        << "\n  from: " << *i << (hasOperation(i) ? "" : " (unregistered)")
+        << "\n  to:   " << *j << (hasOperation(j) ? "" : " (unregistered)");
+    return failure();
+  }
+  return success();
+}
+
+LogicalResult Scheduler::checkOpr(OperatorType *opr) {
+  if (!hasLatency(opr)) {
+    containingOp->emitError()
+        << "Operator type '" << opr->getName() << "' has no latency";
+    return failure();
+  }
+  return success();
+}
+
+LogicalResult Scheduler::checkProb() { return success(); }
+
+/// Check overall problem by delegating to the component-specific checkers.
+LogicalResult Scheduler::check() {
+  for (auto *op : getOperations()) {
```

Comment:
nit: these for loops don't need the curly braces.

Reply:
Done.